### PR TITLE
fix(vuesim): encode BMP and TIFF properly, drop GIF from render image dialog

### LINKED ISSUE #1305: 🐞 Bug: Render Image dialog offers BMP, GIF and TIFF but always saves a PNG
### Is there an existing issue for this?

- [x] I have searched the existing issues.

### Describe the bug

The Render Image dialog offers six formats. Three of them do not work. Pick BMP, GIF or TIFF and you get a PNG file with the wrong extension on it.

The dialog ends up calling `canvas.toDataURL('image/gif')`. A canvas can only encode PNG, JPEG and WEBP. When the type is not supported the spec says fall back to PNG, quietly, with no error. `downloadAsImg` in `src/simulator/src/data/save.js` line 58 then names the file from what you picked rather than from what came back.

I picked each format in turn and read the first bytes of the file that landed:

| picked | saved as | really is | bytes |
|---|---|---|---|
| PNG | Main.png | PNG | 23858 |
| JPEG | Main.jpeg | JPEG | 7895 |
| BMP | Main.bmp | PNG | 23858 |
| GIF | Main.gif | PNG | 23858 |
| TIFF | Main.tiff | PNG | 23858 |

The last three are byte for byte the same file as the PNG. SVG is fine, it goes down a separate path and is built by canvas2svg.

The format list lives in `src/components/DialogBox/SaveImage.vue` line 123, and the same line is in the v1 copy.

### Steps to Reproduce

1. Open the simulator and drop any element on the canvas.
2. Tools, then Download Image.
3. Pick GIF, then click Render Circuit Image.
4. Run `file Main.gif`, or open it in a hex editor. It starts with the PNG signature.

### Expected Behavior

Picking GIF should not hand you a PNG called `.gif`. Either the dialog offers only what the app can really produce, or the file gets named after what was actually made.

### Screenshots

_No response_

### Is the faced issue/bug related to the Vue simulator?

Yes

### Vue simulator version

v0, v1

### Used Vue simulator with or without backend?

Without Backend

### Is the bug present only on the dev server, the build, or both?

Both

### Device Information

```markdown
- **OS**: Windows 11
- **Browser**: Chrome
- **Version**: 152.0.7977.84
```

### Additional Context

Checked on the dev server and on `npm run build` served with `vite preview`. Same result both times. For reference, `canvas.toDataURL` in Chrome returns a `data:image/png` URL for bmp, gif and tiff, and the right prefix for png, jpeg and webp.

### Are you working on this issue?

Yes

### PR DESCRIPTION
**Pull request by iamtanishqjain**
Fixes #1305

#### Describe the changes you have made in this PR -

Updated after @ShinichiShi's review. Instead of dropping BMP and TIFF, they are now written properly from the canvas pixels. GIF is gone, since a canvas snapshot is a single frame and a real animated GIF needs frame capture and a duration control, which is a feature rather than a fix.

New file `src/simulator/src/data/imageFormats.ts` holds the two encoders, uncompressed 24 bit BMP and uncompressed baseline RGB TIFF. `downloadAsImg` uses them for those two types and leaves PNG, JPEG and SVG on the paths they already used.

Downloading each format from the dialog now gives:

| picked | saved as | really is | opens as |
|---|---|---|---|
| PNG | Main.png | PNG | |
| JPEG | Main.jpeg | JPEG | |
| BMP | Main.bmp | BMP | 1280x772 |
| TIFF | Main.tiff | TIFF | 1280x772 |

I decoded both with bmp-js and UTIF to check the pixels survive the trip, and rendered the downloaded BMP back in a browser:

bmp opened again

No new dependencies. The BMP packages on npm all want a Node `Buffer`, and UTIF would cover TIFF but pulls pako in with it. Both formats are simple enough uncompressed that writing them costs about fifty lines each. Say the word and I will swap TIFF over to UTIF.

Tests. `data.spec.js` already covers this directory, so the cases went there in both versions rather than in a new file. One checks which formats the dialog offers, two encode a 2x2 image and check the headers that come out, and one covers a 1x1 image because TIFF wants every offset on an even byte and an odd pixel count used to break that.

Local runs against main at 774003e:

- `npm run test:src:ci` 67 on main, 71 here, all passing
- `npm run test:v1:ci` 79 on main, 83 here, all passing
- `npm run lint:ci` 22 warnings and 0 errors on both, unchanged
- `npm run lint:v1:ci` clean, and that one runs with `--deny-warnings`
- `npm run fmt:check` clean
- `npm run build` fine for v0 and v1

Those 22 warnings are already on main and none sit in files I touched.

### Screenshots of the UI changes (If any) -

Before:

before

After:

after

---

## Code Understanding and AI Usage

**Did you use AI assistance (ChatGPT, Claude, Copilot, etc.) to write any part of this code?**
- [ ] No, I wrote all the code myself
- [x] Yes, I used AI assistance (continue below)

**If you used AI assistance:**
- [x] I have reviewed every single line of the AI-generated code
- [x] I can explain the purpose and logic of each function/component I added
- [x] I have tested edge cases and understand how the code handles them
- [x] I have modified the AI output to follow this project's coding standards and conventions

**Explain your implementation approach:**

`encodeBmp` walks the rows backwards because BMP stores them bottom up, writes each pixel as BGR rather than RGB, and pads every row to a multiple of four bytes. That padding is why the row size is worked out separately from the width.

`encodeTiff` lays the file out as header, pixels, then the small values that do not fit in a tag, then the tag table last. Three tags need somewhere to point at, the three bytes per sample and the two resolution fractions, which is why their offsets are worked out up front. The entries are in ascending tag order because the format requires it.

`encodeCanvas` is the only part that touches the DOM, so the two encoders stay plain functions over a pixel array and can be tested without a canvas.

Alpha is dropped in both. `generateImage` already forces the transparent option off for anything that is not PNG, so there is nothing to lose.

---

## Checklist before requesting a review
- [x] I have added proper PR title and linked to the issue
- [x] I have performed a self-review of my code
- [x] **I can explain the purpose of every function, class, and logic block I added**
- [x] I understand why my changes work and have tested them thoroughly
- [x] I have considered potential edge cases and how my code handles them
- [x] If it is a core feature, I have added thorough tests
- [x] My code follows the project's style guidelines and conventions

---

Note: Please check **Allow edits from maintainers** if you would like us to assist in the PR.

## Summary by CodeRabbit

- **New Features**
  - Added BMP and TIFF export options to the Save Image dialog.
  - BMP and TIFF images can now be downloaded directly from the simulator.

- **Changes**
  - GIF export is no longer available.
  - The Save Image dialog now offers PNG, JPEG, SVG, BMP, and TIFF formats.

- **Tests**
  - Added coverage for available export formats and BMP/TIFF image generation.

### CODE REVIEW

**Comment by tachyons** [comment]
Other formats worked in legacy simulator, if it is not working in vue simulator, then that is a bug

**Comment by iamtanishqjain** [comment]
Hi @tachyons, I saw this was closed and then reopened, so just checking in. Is the PR fine as it is, or would you rather I keep the three formats and name the downloaded file after what actually gets produced?

**Comment by ShinichiShi** [comment]
according to me, it would be great to add the support for having these formats exactly as they are meant to be rather than as png (the concerned packages of tiff and bmp need to be added)
but for a gif it records a single frame of circuit only, i guess that could be removed, else it should have a ui to support the time duration and some more frames of circuits.
cc: @tachyons

**Comment by iamtanishqjain** [comment]
thanks, that makes sense. pushed it, bmp and tiff are now written from the canvas pixels instead of toDataURL, so they actually come out as bmp and tiff. dropped gif for the single frame reason you said, an animated one needs frame capture and a duration control which feels like its own feature.

didn't end up adding packages though, the bmp ones on npm all want a node Buffer and utif brings pako along, and both are simple enough uncompressed to write in about fifty lines each. happy to switch tiff over to utif if you'd rather.

also ptal at #1303 when you get a chance, naman requested your review there.

File: src/simulator/src/data/imageFormats.ts [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+  return out;
+}
+
+/**
+ * Encode pixels as an uncompressed baseline RGB TIFF.
+ */
+export function encodeTiff(rgba: Uint8ClampedArray, width: number, height: number): Uint8Array {
+  const pixelBytes = width * height * 3;
+  const bitsAt = 8 + pixelBytes;
```
[the remark this answers is not in the document]
Comment (iamtanishqjain) [reply src/simulator/src/data/imageFormats.ts]: good catch, that was real. pixelBytes is odd whenever width*height is odd, so every offset after it came out odd. my 2x2 test and the 1280x772 export were both even so nothing caught it.

padded the pixel block to an even length, and added a 1x1 case to the spec so it stays fixed. checked 1x1, 3x1 and 5x3, ifd offset is even now and the pixels still round trip through UTIF.

[the remark this answers is not in the document]
Comment (iamtanishqjain) [reply src/simulator/src/data/imageFormats.ts]: done

**Comment by iamtanishqjain** [comment]
@tachyons the five workflow runs here are still sitting at action_required, so the required checks have never started. Could you approve them when you get a chance? Netlify and CodeRabbit are green.

@ShinichiShi I pushed the real BMP and TIFF encoders after your review and dropped GIF for the single frame reason you gave. Would you mind another look when you have time?
